gradle-plugin/src/gradleTest/java/com/squareup/anvil/plugin/IncrementalTest.kt

```diff
@@ -0,0 +1,545 @@
+      }
+      gradlePropertiesFile(
+        """
+        com.squareup.anvil.trackSourceFiles=true
+        """.trimIndent(),
+      )
+    }
+
+    shouldSucceed("compileJava")
+
+    rootAnvilMainGenerated.injectClassFactory.shouldExist()
+    rootAnvilMainGenerated.resolve("com/squareup/test/OtherClass_Factory.kt").shouldExist()
+
+    rootProject {
+      dir("src/main/java") {
+        kotlinFile(
+          otherClassPath,
+          """
+           package com.squareup.test
+     
+           class OtherClass
+          """.trimIndent(),
+        )
+      }
+    }
+
+    shouldSucceed("compileJava") {
+      task(":compileKotlin")?.outcome shouldBe TaskOutcome.SUCCESS
+    }
+
+    rootAnvilMainGenerated.injectClassFactory.shouldExist()
+    rootAnvilMainGenerated.resolve("com/squareup/test/OtherClass_Factory.kt").shouldNotExist()
+  }
+
+  @TestFactory
+  fun `compilation re-runs when a @ContributesBinding type supertype changes`() = testFactory {
+
+    val otherClassPath = "com/squareup/test/OtherClass.kt"
+
+    rootProject {
+      dir("src/main/java") {
+
+        kotlinFile(
+          otherClassPath,
+          """
+          package com.squareup.test
+
+          import com.squareup.anvil.annotations.ContributesBinding
+          import javax.inject.Inject
+
+          @ContributesBinding(Any::class)
+          class OtherClass @Inject constructor() : TypeA
+
+          interface TypeA
+          interface TypeB
+
+          class Consumer @Inject constructor(
+            private val dep: TypeA
+          )
+          """.trimIndent(),
+        )
+        injectClass()
+      }
+      gradlePropertiesFile(
+        """
+        com.squareup.anvil.trackSourceFiles=true
+        """.trimIndent(),
+      )
+    }
+
+    shouldSucceed("compileJava")
+
+    rootAnvilMainGenerated.injectClassFactory.shouldExist()
+    rootAnvilMainGenerated.resolve("com/squareup/test/OtherClass_Factory.kt").shouldExist()
+    rootAnvilMainGenerated.resolve(
+      "anvil/hint/binding/com/squareup/test/OtherClass.kt",
+    ).shouldExist()
+
+    rootProject.dir("src/main/java") {
+      kotlinFile(
+        otherClassPath,
+        """
+        package com.squareup.test
+
+        import com.squareup.anvil.annotations.ContributesBinding
+        import javax.inject.Inject
+
+        @ContributesBinding(Any::class)
+        class OtherClass @Inject constructor() : TypeB
+
+        interface TypeA
+        interface TypeB
+
+        class Consumer @Inject constructor(
+          private val dep: TypeA
+        )
+        """.trimIndent(),
+      )
+    }
+
+    shouldSucceed("compileJava") {
+      task(":compileKotlin")?.outcome shouldBe TaskOutcome.SUCCESS
+    }
+
+    rootAnvilMainGenerated.injectClassFactory.shouldExist()
+    rootAnvilMainGenerated.resolve("com/squareup/test/OtherClass_Factory.kt").shouldExist()
+    rootAnvilMainGenerated.resolve(
+      "anvil/hint/binding/com/squareup/test/OtherClass.kt",
+    ).shouldExist()
```

Comment:
These would all still be true even without the second compilation right? It feels like we want to utilize a different assert here based on the generated content and possibly changes to the updated source to verify that the supertype change was picked up in the recompilation. My concern right now would be that `compileKotlin` could run successfully but silently not pick up the source file difference and these assertions would still pass

Reply:
There isn't actually any difference in any of the generated output between those two compilations.

The output change only occurs when a merge happens, so I added an `AppComponent` with `@MergeComponent(Any::class)`.

Reply:
Nice, updated assertions are perfect